packages/salesforcedx-utils-vscode/src/commands/sfdxCommandlet.ts

```diff
@@ -0,0 +1,159 @@
+/*
+ * Copyright (c) 2017, salesforce.com, inc.
+ * All rights reserved.
+ * Licensed under the BSD 3-Clause license.
+ * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
+ */
+import * as vscode from 'vscode';
+import { CliCommandExecutor, Command, CommandExecution } from '../cli';
+import {
+  Measurements,
+  Properties,
+  TelemetryData,
+  TelemetryService
+} from '../index';
+import {
+  ContinueResponse,
+  ParametersGatherer,
+  PostconditionChecker,
+  PreconditionChecker
+} from '../types';
+import { getRootWorkspacePath } from '../workspaces';
+// import { ChannelService } from './channelService';
+import { notificationService, ProgressNotification } from './index';
+import { EmptyPostChecker } from './postconditionCheckers';
+
+export interface FlagParameter<T> {
+  flag: T;
+}
+
+export interface CommandletExecutor<T> {
+  execute(response: ContinueResponse<T>): void;
+  readonly onDidFinishExecution?: vscode.Event<[number, number]>;
+}
+
+export abstract class SfdxCommandletExecutor<T>
+  implements CommandletExecutor<T> {
+  protected showChannelOutput = true;
+  protected executionCwd = getRootWorkspacePath();
+  protected onDidFinishExecutionEventEmitter = new vscode.EventEmitter<
+    [number, number]
+  >();
+  public readonly onDidFinishExecution: vscode.Event<[number, number]> = this
+    .onDidFinishExecutionEventEmitter.event;
+
+  protected attachExecution(
+    execution: CommandExecution,
+    cancellationTokenSource: vscode.CancellationTokenSource,
+    cancellationToken: vscode.CancellationToken
+  ) {
+    /* channelService.streamCommandOutput(execution);
+
+    if (this.showChannelOutput) {
+      channelService.showChannelOutput();
+    } */
+
+    notificationService.reportCommandExecutionStatus(
+      execution,
+      cancellationToken
+    );
+    ProgressNotification.show(execution, cancellationTokenSource);
+  }
+
+  public logMetric(
+    logName: string | undefined,
+    hrstart: [number, number],
+    properties?: Properties,
+    measurements?: Measurements
+  ) {
+    TelemetryService.getInstance()
```

Comment:
Not sure but I think we'll have to eventually change this or move it to the extension code. If every extension has it's own telemetry instance I think the current way we handle getting the telemetry instance might have to change to be more like what we have for channel service.

Reply:
That could be a story for another day, but I won't be hurrying to address it in this PR, since it requires more thoughtfulness in the design of it. It could be that we pass ChannelService and TelemetryService instances into SfdxCommandlet.

As of current state, as long as TelemetryService is initialized correctly in the extension, telemetry will be logged.